flink/src/test/java/org/apache/iceberg/flink/TestFlinkCatalogTablePartitions.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.flink;
+
+import java.util.List;
+import org.apache.flink.table.catalog.CatalogPartitionSpec;
+import org.apache.flink.table.catalog.ObjectPath;
+import org.apache.flink.table.catalog.exceptions.TableNotExistException;
+import org.apache.flink.table.catalog.exceptions.TableNotPartitionedException;
+import org.apache.iceberg.AssertHelpers;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+public class TestFlinkCatalogTablePartitions extends FlinkCatalogTestBase {
+
+  private String tableName = "partition_table";
+
+  public TestFlinkCatalogTablePartitions(String catalogName, String[] baseNamespace) {
+    super(catalogName, baseNamespace);
+  }
+
+  @Before
+  public void before() {
+    super.before();
+    sql("CREATE DATABASE %s", flinkDatabase);
+    sql("USE CATALOG %s", catalogName);
+    sql("USE %s", DATABASE);
+  }
+
+  @After
+  public void cleanNamespaces() {
+    sql("DROP TABLE IF EXISTS %s.%s", flinkDatabase, tableName);
+    sql("DROP DATABASE IF EXISTS %s", flinkDatabase);
+    super.clean();
+  }
+
+  @Test
+  public void testListPartitionsWithUnpartitionedTable() throws TableNotExistException, TableNotPartitionedException {
+    sql("CREATE TABLE %s (id INT, data VARCHAR)", tableName);
+    sql("INSERT INTO %s SELECT 1,'a'", tableName);
+
+    ObjectPath objectPath = new ObjectPath(DATABASE, tableName);
+    FlinkCatalog flinkCatalog = (FlinkCatalog) getTableEnv().getCatalog(catalogName).get();
+    flinkCatalog.loadIcebergTable(objectPath).refresh();
```

Comment:
Do we need this `refresh` sentence ?  For the newly loaded iceberg table will refresh the latest metadata automatically ,  so I don't think we need this.

Reply:
For unpartitioned tables, refresh is really not needed, but for partitioned tables, if we  do not add refresh , data will not be queried.

When I wrote the test case, the result of the test is that if use` validationCatalog.loadTable(..).refresh() `after `insert sql `in the test case, it is also invalid.
So I add `flinkCatalog.loadIcebergTable(objectPath).refresh(); ` to refresh the table.

In addition, if I write the following flink program to get the partition, it is no problem. It does not need to be refreshed, which proves that the main program is no problem. I guess there may be a issue in a certain part of the test case.

```
	StreamTableEnvironment tenv = StreamTableEnvironment.create(env);
		String sql = " CREATE CATALOG iceberg WITH (\n" +
		             "  'type'='iceberg',\n" +
		             "   'catalog-type'='hive',\n" +
		             "   'hive-conf-dir'='/Users/user/work/hive/conf/',\n" +
		             "    'uri'='thrift://localhost:9083'\n" +
		             " )";
		tenv.executeSql(sql);
		tenv.useCatalog("iceberg");
		FlinkCatalog catalog = (FlinkCatalog) tenv.getCatalog("iceberg").get();
		List<CatalogPartitionSpec> list = catalog.listPartitions(new ObjectPath("iceberg_db", "iceberg_table"));

```

Reply:
I think that's because the `FlinkCatalogFactory` have created a cached iceberg catalog.  Even if the table has written few new partitions, the stale table won't catch the changes.

Reply:
Instead of refreshing the iceberg table in this unit test,   I think we need to refresh the loaded iceberg table in [loadIcebergTable](https://github.com/apache/iceberg/blob/master/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java#L306) if the `cacheEnabled ` is true.   In [SparkTable](https://github.com/apache/iceberg/blob/master/spark3/src/main/java/org/apache/iceberg/spark/source/SparkTable.java#L73),  I saw we had a switch named `refreshEagerly`  which have the similar effect. 

btw,  we may need more unit tests to cover both cache enabled and disabled cases for `listPartitions`.

Reply:
I added the cache enabled test case